include/zephyr/zbus/zbus.h

```diff
@@ -0,0 +1,653 @@
+/*
+ * Copyright (c) 2022 Rodrigo Peixoto <rodrigopex@gmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZBUS_H_
+#define ZEPHYR_INCLUDE_ZBUS_H_
+
+#include <string.h>
+
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Zbus API
+ * @defgroup zbus_apis Zbus APIs
+ * @{
+ */
+
+/**
+ * @brief Type used to represent a channel.
+ *
+ * Every channel has a zbus_channel structure associated used to control the channel
+ * access and usage.
+ */
+struct zbus_channel {
+#if defined(CONFIG_ZBUS_CHANNEL_NAME) || defined(__DOXYGEN__)
+	/** Channel name. */
+	const char *const name;
+#endif
+	/** Message size. Represents the channel's message size. */
+	const uint16_t message_size;
+
+	/** User data available to extend zbus features. The channel must be claimed before
+	 * using this field.
+	 */
+	void *const user_data;
+
+	/** Message reference. Represents the message's reference that points to the actual
+	 * shared memory region.
+	 */
+	void *const message;
+
+	/** Message validator. Stores the reference to the function to check the message
+	 * validity before actually performing the publishing. No invalid messages can be
+	 * published. Every message is valid when this field is empty.
+	 */
+	bool (*const validator)(const void *msg, size_t msg_size);
+
+	/** Access control mutex. Points to the mutex used to avoid race conditions
+	 * for accessing the channel.
+	 */
+	struct k_mutex *mutex;
```

Comment:
The observers field is already const. The other two cannot be const because of these errors:

```console
/.../zephyrproject/zephyr/subsys/zbus/zbus_runtime_observers.c:41:44: warning: passing argument 1 of ‘k_mutex_unlock’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   41 |                         k_mutex_unlock(chan->mutex);
      |                                        ~~~~^~~~~~~
```

```console
/home/rodrigopex/Cloud/Projects/zephyrproject/zephyr/subsys/zbus/zbus_runtime_observers.c:39:47: warning: passing argument 1 of ‘sys_slist_peek_head’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   39 |         SYS_SLIST_FOR_EACH_CONTAINER_SAFE(chan->runtime_observers, obs_nd, tmp, node) {
      |                                           ~~~~^~~~~~~~~~~~~~~~~~~
/.../zephyrproject/zephyr/include/zephyr/sys/util.h:211:29: note: in definition of macro ‘CONTAINER_OF’
  211 |         ((type *)(((char *)(ptr)) - offsetof(type, field)))
      |                             ^~~
/.../zephyrproject/zephyr/include/zephyr/sys/list_gen.h:35:9: note: in expansion of macro ‘Z_GENLIST_CONTAINER’
   35 |         Z_GENLIST_CONTAINER(sys_ ## __lname ## _peek_head(__l), __cn, __n)
      |         ^~~~~~~~~~~~~~~~~~~
/.../zephyrproject/zephyr/include/zephyr/sys/list_gen.h:52:21: note: in expansion of macro ‘Z_GENLIST_PEEK_HEAD_CONTAINER’
   52 |         for (__cn = Z_GENLIST_PEEK_HEAD_CONTAINER(__lname, __l, __cn, __n),   \
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/.../zephyrproject/zephyr/include/zephyr/sys/slist.h:181:9: note: in expansion of macro ‘Z_GENLIST_FOR_EACH_CONTAINER_SAFE’
  181 |         Z_GENLIST_FOR_EACH_CONTAINER_SAFE(slist, __sl, __cn, __cns, __n)
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/.../zephyrproject/zephyr/subsys/zbus/zbus_runtime_observers.c:39:9: note: in expansion of macro ‘SYS_SLIST_FOR_EACH_CONTAINER_SAFE’
   39 |         SYS_SLIST_FOR_EACH_CONTAINER_SAFE(chan->runtime_observers, obs_nd, tmp, node) {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /.../zephyrproject/zephyr/include/zephyr/kernel_includes.h:25,
                 from /.../zephyrproject/zephyr/include/zephyr/kernel.h:17,
                 from /.../zephyrproject/zephyr/subsys/zbus/zbus_runtime_observers.c:5:
/.../zephyrproject/zephyr/include/zephyr/sys/slist.h:231:61: note: expected ‘sys_slist_t *’ {aka ‘struct _slist *’} but argument is of type ‘const sys_slist_t *’ {aka ‘const struct _slist *’}
  231 | static inline sys_snode_t *sys_slist_peek_head(sys_slist_t *list)
      |                                                ~~~~~~~~~~~~~^~~~
```

Reply:
> The observers field is already const.

No, it's not. And also the array it points to is not, though it should be. Currently only the structures pointed to by items of the observers array are `const`. That's what you should use:
```
	const struct zbus_observer *const *const observers;
```

> The other two cannot be const because of these warnings:

You get the warnings because you are not making the fields `const` but the values they point to (which indeed should not be `const`, because they are supposed to be modified).
`const struct k_mutex *mutex;` has different meaning than `struct k_mutex *const mutex;`.